### PR TITLE
Update to Checker dataflow 3.13.0

### PR DESCRIPTION
**Pull request by msridhar**
Helps out with #462.  Should land after #469 

I needed to adjust to some API changes, and re-implement some deprecated functionality that we want to keep.  Also, it looks like a try-finally false negative got fixed!  I filed https://github.com/typetools/checker-framework/issues/4612 on the need to add the explicit `plume-util` dependence.

Before landing, this should be tested internally, both for perf and also to check if new errors are reported.  Nothing urgent on my end.

### CODE REVIEW

**Comment by msridhar** [comment]
Unfortunately clicking the links in that @coveralls comment doesn't show me the actual lines where the coverage changed...

**Comment by msridhar** [comment]
Let's hold off on landing this.  There are some issues with the new dependency I'd like to think through / work out first.

**Comment by msridhar** [comment]
So I think this one is good for review / internal testing.  The one risk here is that currently the `dataflow-shaded` artifact used by Error Prone includes some unshaded classes from the `plume-util` library that could hypothetically conflict with `plume-util` classes that this latest version of `dataflow` depends on.  I've run NullAway tests on all supported versions of Error Prone and they work, so things are fine for now.  Also, the issue with `dataflow-shaded` is getting fixed (see https://github.com/typetools/checker-framework/issues/4615) so this will most likely not be a problem moving forward.  I think the worst case is that some future update of `dataflow` that we want to do would be blocked due to a conflict with Error Prone, but that's independent of this PR and could likely be addressed by asking Error Prone to update their `dataflow-shaded` dependence.    

So, TL;DR most likely we are fine 🙂 Mostly just wanted to write the above to document it somewhere.

**Comment by lazaroclapp** [comment]
Ok, will go over it tomorrow and try to test it internally asap on both our repos and let you know!

**Comment by lazaroclapp** [comment]
Seeing the following error internally:

```
     error-prone version: 2.6.0
     BugPattern: NullAway
     Stack Trace:
     com.google.common.util.concurrent.ExecutionError: java.lang.NoClassDefFoundError: org/checkerframework/dataflow/qual/TerminatesExecution
        at com.google.common.cache.LocalCache$Segment.get(LocalCache.java:2049)
        at com.google.common.cache.LocalCache.get(LocalCache.java:3951)
        at com.google.common.cache.LocalCache.getOrLoad(LocalCache.java:3974)
        at com.google.common.cache.LocalCache$LocalLoadingCache.get(LocalCache.java:4958)
        at com.google.common.cache.LocalCache$LocalLoadingCache.getUnchecked(LocalCache.java:4964)
        at com.uber.nullaway.dataflow.DataFlow.dataflow(DataFlow.java:146)
        at com.uber.nullaway.dataflow.DataFlow.finalResult(DataFlow.java:222)
        at com.uber.nullaway.dataflow.AccessPathNullnessAnalysis.getNonnullFieldsOfReceiverAtExit(AccessPathNullnessAnalysis.java:143)
        at com.uber.nullaway.NullAway.guaranteedNonNullForConstructor(NullAway.java:1588)
        at com.uber.nullaway.NullAway.checkConstructorInitialization(NullAway.java:1565)
        at com.uber.nullaway.NullAway.checkFieldInitialization(NullAway.java:1424)
        at com.uber.nullaway.NullAway.matchClass(NullAway.java:1154)
        at com.google.errorprone.scanner.ErrorProneScanner.processMatchers(ErrorProneScanner.java:450)
        at com.google.errorprone.scanner.ErrorProneScanner.visitClass(ErrorProneScanner.java:548)
        at com.google.errorprone.scanner.ErrorProneScanner.visitClass(ErrorProneScanner.java:151)
        at com.sun.tools.javac.tree.JCTree$JCClassDecl.accept(JCTree.java:808)
        at com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:82)
        at com.google.errorprone.scanner.Scanner.scan(Scanner.java:74)
        at com.google.errorprone.scanner.Scanner.scan(Scanner.java:48)
        at com.sun.source.util.TreeScanner.scan(TreeScanner.java:105)
        at com.sun.source.util.TreeScanner.scanAndReduce(TreeScanner.java:113)
        at com.sun.source.util.TreeScanner.visitCompilationUnit(TreeScanner.java:144)
        at com.google.errorprone.scanner.ErrorProneScanner.visitCompilationUnit(ErrorProneScanner.java:561)
        at com.google.errorprone.scanner.ErrorProneScanner.visitCompilationUnit(ErrorProneScanner.java:151)
        at com.sun.tools.javac.tree.JCTree$JCCompilationUnit.accept(JCTree.java:591)
        at com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:56)
        at com.google.errorprone.scanner.Scanner.scan(Scanner.java:58)
        at com.google.errorprone.scanner.ErrorProneScannerTransformer.apply(ErrorProneScannerTransformer.java:43)
        at com.google.errorprone.ErrorProneAnalyzer.finished(ErrorProneAnalyzer.java:152)
        at com.sun.tools.javac.api.MultiTaskListener.finished(MultiTaskListener.java:120)
        at com.sun.tools.javac.main.JavaCompiler.flow(JavaCompiler.java:1404)
        at com.sun.tools.javac.main.JavaCompiler.flow(JavaCompiler.java:1353)
        at com.sun.tools.javac.main.JavaCompiler.compile(JavaCompiler.java:946)
        at com.sun.tools.javac.api.JavacTaskImpl.lambda$doCall$0(JavacTaskImpl.java:100)
        at com.sun.tools.javac.api.JavacTaskImpl.handleExceptions(JavacTaskImpl.java:142)
        at com.sun.tools.javac.api.JavacTaskImpl.doCall(JavacTaskImpl.java:96)
        at com.sun.tools.javac.api.JavacTaskImpl.call(JavacTaskImpl.java:90)
        at com.facebook.buck.jvm.java.plugin.adapter.JavacTaskWrapper.call(JavacTaskWrapper.java:100)
        at com.facebook.buck.jvm.java.plugin.adapter.BuckJavacTask.call(BuckJavacTask.java:52)
        at com.facebook.buck.jvm.java.plugin.adapter.BuckJavacTaskProxyImpl.call(BuckJavacTaskProxyImpl.java:134)
        at com.facebook.buck.jvm.java.Jsr199JavacInvocation$CompilerWorker.lambda$startCompiler$2(Jsr199JavacInvocation.java:424)
        at com.google.common.util.concurrent.TrustedListenableFutureTask$TrustedFutureInterruptibleTask.runInterruptibly(TrustedListenableFutureTask.java:127)
        at com.google.common.util.concurrent.InterruptibleTask.run(InterruptibleTask.java:57)
        at com.google.common.util.concurrent.TrustedListenableFutureTask.run(TrustedListenableFutureTask.java:80)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at java.lang.Thread.run(Thread.java:748)
  Caused by: java.lang.NoClassDefFoundError: org/checkerframework/dataflow/qual/TerminatesExecution
        at org.checkerframework.dataflow.cfg.builder.CFGTranslationPhaseOne.visitMethodInvocation(CFGTranslationPhaseOne.java:1350)
        at org.checkerframework.dataflow.cfg.builder.CFGTranslationPhaseOne.visitMethodInvocation(CFGTranslationPhaseOne.java:198)
        at com.sun.tools.javac.tree.JCTree$JCMethodInvocation.accept(JCTree.java:1644)
        at com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:82)
        at org.checkerframework.dataflow.cfg.builder.CFGTranslationPhaseOne.visitExpressionStatement(CFGTranslationPhaseOne.java:2314)
        at org.checkerframework.dataflow.cfg.builder.CFGTranslationPhaseOne.visitExpressionStatement(CFGTranslationPhaseOne.java:198)
        at com.sun.tools.javac.tree.JCTree$JCExpressionStatement.accept(JCTree.java:1454)
        at com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:82)
        at org.checkerframework.dataflow.cfg.builder.CFGTranslationPhaseOne.visitBlock(CFGTranslationPhaseOne.java:2062)
        at org.checkerframework.dataflow.cfg.builder.CFGTranslationPhaseOne.visitBlock(CFGTranslationPhaseOne.java:198)
        at com.sun.tools.javac.tree.JCTree$JCBlock.accept(JCTree.java:1026)
        at com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:56)
        at org.checkerframework.dataflow.cfg.builder.CFGTranslationPhaseOne.process(CFGTranslationPhaseOne.java:428)
        at org.checkerframework.dataflow.cfg.builder.CFGBuilder.build(CFGBuilder.java:91)
        at com.uber.nullaway.dataflow.DataFlow$2.load(DataFlow.java:130)
        at com.uber.nullaway.dataflow.DataFlow$2.load(DataFlow.java:97)
        at com.google.common.cache.LocalCache$LoadingValueReference.loadFuture(LocalCache.java:3529)
        at com.google.common.cache.LocalCache$Segment.loadSync(LocalCache.java:2278)
        at com.google.common.cache.LocalCache$Segment.lockedGetOrLoad(LocalCache.java:2155)
        at com.google.common.cache.LocalCache$Segment.get(LocalCache.java:2045)
        ... 46 more
  Caused by: java.lang.ClassNotFoundException: org.checkerframework.dataflow.qual.TerminatesExecution
        at java.net.URLClassLoader.findClass(URLClassLoader.java:382)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:418)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:351)
        ... 66 more
       
```

Which seems correct in that `org.checkerframework.dataflow.qual.TerminatesExecution` is not part of [org.checkerframework:dataflow:jar:3.13.0](https://repo1.maven.org/maven2/org/checkerframework/dataflow/3.13.0/).

See:

```
% unzip -l ~unzip -l ~/Downloads/dataflow-3.13.0.jar | grep TerminatesExecution
% [no output]
```

The corresponding `org.checkerframework.shaded.dataflow.qual.TerminatesExecution` is however part of [org.checkerframework:dataflow-shaded:jar:3.13.0](https://mvnrepository.com/artifact/org.checkerframework/dataflow-shaded/3.13.0)

```
% unzip -l ~/Downloads/dataflow-shaded-3.13.0.jar | grep TerminatesExecution 
      492  05-03-2021 12:33   org/checkerframework/shaded/dataflow/qual/TerminatesExecution.class
```

However, keep in mind that EP 2.6.0 pulls in `dataflow-shaded:3.11.0`

**Comment by msridhar** [comment]
Yikes, that looks like a Checker Framework packaging issue.  Will look into it.  @lazaroclapp any chance of producing an open-source test case that repros the error?

**Comment by msridhar** [comment]
@lazaroclapp that annotation is inside the [checker-qual 3.13.0 artifact](https://search.maven.org/artifact/org.checkerframework/checker-qual/3.13.0/jar), which is a dependence of dataflow.  Do you need to update the version of checker-qual used in the internal build?

**Comment by msridhar** [comment]
Closing in favor of #485
